invenio_rdm_migrator/state.py

```diff
@@ -0,0 +1,248 @@
+            .values(**data)
+        )
+        assert res.rowcount == 1  # the update succeeded
+
+    def _copy_db(cls, src, dest):
+        """Copy source database into destination."""
+        # sqlalchemy abstraction does not have a backup we need a raw connection
+        src_conn = src.raw_connection()
+        dst_conn = dest.raw_connection()
+
+        src_conn.backup(dst_conn.connection)
+
+        src_conn.close()
+        dst_conn.close()
+
+    def _validate(self, table, data):
+        """Validate data for a specific table."""
+        validator = self.validators.get(table.name)
+        if validator:
+            validator.validate(data)
+
+    @classmethod
+    def _initialize_db(cls, metadata):
+        """Initialize state tables and bind to metadata object."""
+        # this registration method is not ideal. It implies a tight coupling between
+        # state and which streams can be run, and there is no independence (i.e. running one
+        # stream means having all state tables defined rather than only those needed).
+        # The tables should be defined per stream (e.g. streams/records/state.py:ParentTable)
+        # and then be injected into the state.
+
+        Table(
+            "parents",
+            metadata,
+            Column("recid", String, primary_key=True),
+            Column("id", UUIDType, nullable=False, unique=True),
+            Column("latest_id", UUIDType),
+            # could use default=0, but it could create uncontrolled scenarios
+            Column("latest_index", Integer),
+            Column("next_draft_id", UUIDType),
+            sqlite_autoincrement=False,
+        )
+
+        Table(
+            "records",
+            metadata,
+            Column("recid", String, primary_key=True),
+            Column("id", UUIDType, nullable=False, unique=True),
+            Column("parent_id", UUIDType, nullable=False),
+            Column("index", Integer, nullable=False),
+            Column("fork_version_id", Integer, nullable=False),
+            Column("pids", JSONType, default="{}"),
+            sqlite_autoincrement=False,
+        )
+
+        Table(
+            "communities",
+            metadata,
+            Column("slug", String, primary_key=True),
+            Column("id", UUIDType, unique=True),
+            sqlite_autoincrement=False,
+        )
+
+        # e.g. key = max_pid_pk, value = 1000000
+        Table(
+            "global",
+            metadata,
+            Column("key", String, primary_key=True),
+            Column("value", Integer, primary_key=True),
+            sqlite_autoincrement=False,
+        )
+
+
+class GLOBAL:
+    """Static class holding a reference to a state query."""
+
+    # implemented this way to avoid having to drill state.global from the
+    # runner to the lower level pid_pk()
+    # ideally there would be some sort of application context/proxies to access this.
+    STATE = None
+
+
+class StateQuery:
```

Comment:
```suggestion
class StateEntity:
```

minor/nit: or something similar. Since it's also used for writes/updates.